spec/support/expectations/equal_namespace_definitions.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require 'rspec/expectations'
+
+RSpec::Matchers.define :have_equal_namespace_definitions_as do |expected|
+  attr_reader :actual, :expected
+
+  match do |actual|
+    return false unless actual.is_a?(String) && expected.is_a?(String)
+
+    expected_xml = parse_fragment(expected)
+    @expected = expected_xml.to_xml
+    actual_xml = parse_fragment(actual)
+    @actual = actual_xml.to_xml
+
+    return false if expected_xml.errors.length.positive? || actual_xml.errors.length.positive?
+    return false unless EquivalentXml.equivalent?(expected_xml, actual_xml)
```

Comment:
Is it necessary to check for XML-equality as well? It feels out of scope for namespace comparison.

Reply:
Not necessarily. However, my implementation explicitly requires XML-equality. This is caused by the way I extract the namespace definitions and how I compare them: Nokogiri is very smart in handling the namespaces and performs a deduplication of namespaces while reading. Hence, I need to use the `namespace_definitions` method on each node to get the namespaces explicitly defined there. As a consequence, I need to traverse both node trees.

With the assumption that the trees are equal, it is enough to traverse one tree and perform the similar operation on the other tree. Otherwise, I would need to traverse both trees independently from each other and then extract the definitions at each node and level. When one of the trees is not as high as the other, I need to check whether the there are further namespaces definitions in this additional height.

This checking would make the matcher much more complex, so that I refrained from it (since it is not used either). As a consequence, I need to check for XML-quality.

Reply:
Makes sense, thanks for the explanation